packages/phone-number-privacy/signer/src/signer/sign.abstract.ts

```diff
@@ -0,0 +1,62 @@
+import {
+  DomainRestrictedSignatureRequest,
+  KEY_VERSION_HEADER,
+  SignMessageRequest,
+  WarningMessage,
+} from '@celo/phone-number-privacy-common'
+import Logger from 'bunyan'
+import { Request } from 'express'
+import { computeBlindedSignature } from '../bls/bls-cryptography-client'
+import { getKeyProvider } from '../key-management/key-provider'
+import { Key } from '../key-management/key-provider-base'
+import { IAction, Session } from './action.interface'
+import { IOAbstract } from './io.abstract'
+
+declare type OdisSignatureRequest = SignMessageRequest | DomainRestrictedSignatureRequest
+
+export abstract class SignAbstract<R extends OdisSignatureRequest> implements IAction<R> {
+  abstract readonly io: IOAbstract<R>
+
+  public abstract perform(session: Session<R>): Promise<void>
+
+  protected async sign(
+    blindedMessage: string,
+    defaultKey: Key,
+    session: Session<R>
+  ): Promise<{ signature: string; key: Key }> {
+    let keyVersion = this.getRequestKeyVersion(session.request, session.logger)
```

Comment:
I was surprise by this code factoring. I feel resolving the key should occur outside this function and have the final `key` as an input rather than an output.